[
    {
        "number": 23,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "requests"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-9wx4-h78v-vm56",
            "cve_id": "CVE-2024-35195",
            "summary": "Requests `Session` object does not verify requests after making first request with verify=False",
            "description": "When making requests through a Requests `Session`, if the first request is made with `verify=False` to disable cert verification, all subsequent requests to the same origin will continue to ignore cert verification regardless of changes to the value of `verify`. This behavior will continue for the lifecycle of the connection in the connection pool.\n\n### Remediation\nAny of these options can be used to remediate the current issue, we highly recommend upgrading as the preferred mitigation.\n\n* Upgrade to `requests>=2.32.0`.\n* For `requests<2.32.0`, avoid setting `verify=False` for the first request to a host while using a Requests Session.\n* For `requests<2.32.0`, call `close()` on `Session` objects to clear existing connections if `verify=False` is used.\n\n### Related Links\n* https://github.com/psf/requests/pull/6655",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-9wx4-h78v-vm56",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-35195",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/psf/requests/security/advisories/GHSA-9wx4-h78v-vm56"
                },
                {
                    "url": "https://github.com/psf/requests/pull/6655"
                },
                {
                    "url": "https://github.com/psf/requests/commit/a58d7f2ffb4d00b46dca2d70a3932a0b37e22fac"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-35195"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/IYLSNK5TL46Q6XPRVMHVWS63MVJQOK4Q"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/N7WP6EYDSUOCOJYHDK5NX43PYZ4SNHGZ"
                },
                {
                    "url": "https://github.com/advisories/GHSA-9wx4-h78v-vm56"
                }
            ],
            "published_at": "2024-05-20T20:15:00Z",
            "updated_at": "2024-06-10T18:31:04Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "requests"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": "< 2.32.0",
                    "first_patched_version": {
                        "identifier": "2.32.0"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:L/AC:H/PR:H/UI:R/S:U/C:H/I:H/A:N",
                "score": 5.6
            },
            "cwes": [
                {
                    "cwe_id": "CWE-670",
                    "name": "Always-Incorrect Control Flow Implementation"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "requests"
            },
            "severity": "medium",
            "vulnerable_version_range": "< 2.32.0",
            "first_patched_version": {
                "identifier": "2.32.0"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/23",
        "html_url": "https://github.com/username/gradio/security/dependabot/23",
        "created_at": "2024-05-24T01:34:14Z",
        "updated_at": "2024-05-24T01:34:14Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 22,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "Jinja2"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-h75v-3vvj-5mfj",
            "cve_id": "CVE-2024-34064",
            "summary": "Jinja vulnerable to HTML attribute injection when passing user input as keys to xmlattr filter",
            "description": "The `xmlattr` filter in affected versions of Jinja accepts keys containing non-attribute characters. XML/HTML attributes cannot contain spaces, `/`, `>`, or `=`, as each would then be interpreted as starting a separate attribute. If an application accepts keys (as opposed to only values) as user input, and renders these in pages that other users see as well, an attacker could use this to inject other attributes and perform XSS. The fix for the previous GHSA-h5c8-rqwp-cp95 CVE-2024-22195 only addressed spaces but not other characters.\n\nAccepting keys as user input is now explicitly considered an unintended use case of the `xmlattr` filter, and code that does so without otherwise validating the input should be flagged as insecure, regardless of Jinja version. Accepting _values_ as user input continues to be safe.",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-h75v-3vvj-5mfj",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-34064",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/pallets/jinja/security/advisories/GHSA-h75v-3vvj-5mfj"
                },
                {
                    "url": "https://github.com/pallets/jinja/commit/0668239dc6b44ef38e7a6c9f91f312fd4ca581cb"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-34064"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/567XIGSZMABG6TSMYWD7MIYNJSUQQRUC"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/GCLF44KY43BSVMTE6S53B4V5WP3FRRSE"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/SSCBHIL6BYKR5NRCBXP4XMP2CEEKGFVS"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/ZALNWE3TXPPHVPSI3AZ5CTMSTAVN5UMS"
                },
                {
                    "url": "https://github.com/advisories/GHSA-h75v-3vvj-5mfj"
                }
            ],
            "published_at": "2024-05-06T14:20:59Z",
            "updated_at": "2024-06-10T18:30:58Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "Jinja2"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": "< 3.1.4",
                    "first_patched_version": {
                        "identifier": "3.1.4"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:U/C:L/I:L/A:N",
                "score": 5.4
            },
            "cwes": [
                {
                    "cwe_id": "CWE-79",
                    "name": "Improper Neutralization of Input During Web Page Generation ('Cross-site Scripting')"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "Jinja2"
            },
            "severity": "medium",
            "vulnerable_version_range": "< 3.1.4",
            "first_patched_version": {
                "identifier": "3.1.4"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/22",
        "html_url": "https://github.com/username/gradio/security/dependabot/22",
        "created_at": "2024-05-24T01:34:14Z",
        "updated_at": "2024-05-24T01:34:14Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 21,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "tqdm"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-g7vv-2v7x-gj9p",
            "cve_id": "CVE-2024-34062",
            "summary": "tqdm CLI arguments injection attack",
            "description": "### Impact\nAny optional non-boolean CLI arguments (e.g. `--delim`, `--buf-size`, `--manpath`) are passed through python's `eval`, allowing arbitrary code execution. Example:\n\n```sh\npython -m tqdm --manpath=\"\\\" + str(exec(\\\"import os\\nos.system('echo hi && killall python3')\\\")) + \\\"\"\n```\n\n### Patches\nhttps://github.com/tqdm/tqdm/commit/4e613f84ed2ae029559f539464df83fa91feb316 released in `tqdm>=4.66.3`\n\n### Workarounds\nNone\n\n### References\n- https://github.com/tqdm/tqdm/releases/tag/v4.66.3",
            "severity": "low",
            "identifiers": [
                {
                    "value": "GHSA-g7vv-2v7x-gj9p",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-34062",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/tqdm/tqdm/security/advisories/GHSA-g7vv-2v7x-gj9p"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-34062"
                },
                {
                    "url": "https://github.com/tqdm/tqdm/commit/4e613f84ed2ae029559f539464df83fa91feb316"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/PA3GIGHPWAHCTT4UF57LTPZGWHAX3GW6"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/QRECVQCCESHBS3UJOWNXQUIX725TKNY6"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/VA337CYUS4SLRFV2P6MX6MZ2LKFURKJC"
                },
                {
                    "url": "https://github.com/advisories/GHSA-g7vv-2v7x-gj9p"
                }
            ],
            "published_at": "2024-05-03T19:33:28Z",
            "updated_at": "2024-06-10T18:30:58Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "tqdm"
                    },
                    "severity": "low",
                    "vulnerable_version_range": ">= 4.4.0, < 4.66.3",
                    "first_patched_version": {
                        "identifier": "4.66.3"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:L/AC:L/PR:L/UI:R/S:U/C:L/I:L/A:N",
                "score": 3.9
            },
            "cwes": [
                {
                    "cwe_id": "CWE-74",
                    "name": "Improper Neutralization of Special Elements in Output Used by a Downstream Component ('Injection')"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "tqdm"
            },
            "severity": "low",
            "vulnerable_version_range": ">= 4.4.0, < 4.66.3",
            "first_patched_version": {
                "identifier": "4.66.3"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/21",
        "html_url": "https://github.com/username/gradio/security/dependabot/21",
        "created_at": "2024-05-24T01:34:14Z",
        "updated_at": "2024-05-24T01:34:14Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 20,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "idna"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-jjg7-2v4v-x38h",
            "cve_id": "CVE-2024-3651",
            "summary": "Internationalized Domain Names in Applications (IDNA) vulnerable to denial of service from specially crafted inputs to idna.encode",
            "description": "### Impact\nA specially crafted argument to the `idna.encode()` function could consume significant resources. This may lead to a denial-of-service.\n\n### Patches\nThe function has been refined to reject such strings without the associated resource consumption in version 3.7.\n\n### Workarounds\nDomain names cannot exceed 253 characters in length, if this length limit is enforced prior to passing the domain to the `idna.encode()` function it should no longer consume significant resources. This is triggered by arbitrarily large inputs that would not occur in normal usage, but may be passed to the library assuming there is no preliminary input validation by the higher-level application.\n\n### References\n* https://huntr.com/bounties/93d78d07-d791-4b39-a845-cbfabc44aadb",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-jjg7-2v4v-x38h",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-3651",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/kjd/idna/security/advisories/GHSA-jjg7-2v4v-x38h"
                },
                {
                    "url": "https://github.com/advisories/GHSA-jjg7-2v4v-x38h"
                }
            ],
            "published_at": "2024-04-11T21:32:40Z",
            "updated_at": "2024-04-11T21:32:40Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "idna"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": "< 3.7",
                    "first_patched_version": {
                        "identifier": "3.7"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:L/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
                "score": 6.2
            },
            "cwes": [
                {
                    "cwe_id": "CWE-400",
                    "name": "Uncontrolled Resource Consumption"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "idna"
            },
            "severity": "medium",
            "vulnerable_version_range": "< 3.7",
            "first_patched_version": {
                "identifier": "3.7"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/20",
        "html_url": "https://github.com/username/gradio/security/dependabot/20",
        "created_at": "2024-05-24T01:34:14Z",
        "updated_at": "2024-05-24T01:34:14Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 19,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "transformers"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-37q5-v5qm-c9v8",
            "cve_id": "CVE-2024-3568",
            "summary": "Transformers Deserialization of Untrusted Data vulnerability",
            "description": "The huggingface/transformers library is vulnerable to arbitrary code execution through deserialization of untrusted data within the `load_repo_checkpoint()` function of the `TFPreTrainedModel()` class. Attackers can execute arbitrary code and commands by crafting a malicious serialized payload, exploiting the use of `pickle.load()` on data from potentially untrusted sources. This vulnerability allows for remote code execution (RCE) by deceiving victims into loading a seemingly harmless checkpoint during a normal training process, thereby enabling attackers to execute arbitrary code on the targeted machine.",
            "severity": "low",
            "identifiers": [
                {
                    "value": "GHSA-37q5-v5qm-c9v8",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-3568",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-3568"
                },
                {
                    "url": "https://github.com/huggingface/transformers/commit/693667b8ac8138b83f8adb6522ddaf42fa07c125"
                },
                {
                    "url": "https://huntr.com/bounties/b3c36992-5264-4d7f-9906-a996efafba8f"
                },
                {
                    "url": "https://github.com/advisories/GHSA-37q5-v5qm-c9v8"
                }
            ],
            "published_at": "2024-04-10T18:30:48Z",
            "updated_at": "2024-04-10T22:20:57Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "transformers"
                    },
                    "severity": "low",
                    "vulnerable_version_range": "< 4.38.0",
                    "first_patched_version": {
                        "identifier": "4.38.0"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.0/AV:N/AC:H/PR:N/UI:R/S:C/C:N/I:N/A:L",
                "score": 3.4
            },
            "cwes": [
                {
                    "cwe_id": "CWE-502",
                    "name": "Deserialization of Untrusted Data"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "transformers"
            },
            "severity": "low",
            "vulnerable_version_range": "< 4.38.0",
            "first_patched_version": {
                "identifier": "4.38.0"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/19",
        "html_url": "https://github.com/username/gradio/security/dependabot/19",
        "created_at": "2024-05-24T01:34:14Z",
        "updated_at": "2024-05-24T01:34:14Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 18,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "pillow"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-44wm-f244-xhp3",
            "cve_id": "CVE-2024-28219",
            "summary": "Pillow buffer overflow vulnerability",
            "description": "In _imagingcms.c in Pillow before 10.3.0, a buffer overflow exists because strcpy is used instead of strncpy.",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-44wm-f244-xhp3",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-28219",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-28219"
                },
                {
                    "url": "https://pillow.readthedocs.io/en/stable/releasenotes/10.3.0.html#security"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/commit/2a93aba5cfcf6e241ab4f9392c13e3b74032c061"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2024/04/msg00008.html"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce%40lists.fedoraproject.org/message/4XLPUT3VK4GQ6EVY525TT2QNUIXNRU5M"
                },
                {
                    "url": "https://github.com/advisories/GHSA-44wm-f244-xhp3"
                }
            ],
            "published_at": "2024-04-03T03:30:30Z",
            "updated_at": "2024-04-20T05:06:18Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "pillow"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": "< 10.3.0",
                    "first_patched_version": {
                        "identifier": "10.3.0"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:L/AC:H/PR:L/UI:R/S:U/C:H/I:H/A:H",
                "score": 6.7
            },
            "cwes": [
                {
                    "cwe_id": "CWE-120",
                    "name": "Buffer Copy without Checking Size of Input ('Classic Buffer Overflow')"
                },
                {
                    "cwe_id": "CWE-676",
                    "name": "Use of Potentially Dangerous Function"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "pillow"
            },
            "severity": "medium",
            "vulnerable_version_range": "< 10.3.0",
            "first_patched_version": {
                "identifier": "10.3.0"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/18",
        "html_url": "https://github.com/username/gradio/security/dependabot/18",
        "created_at": "2024-05-24T01:34:14Z",
        "updated_at": "2024-05-24T01:34:14Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 17,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "orjson"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-pwr2-4v36-6qpr",
            "cve_id": "CVE-2024-27454",
            "summary": "orjson does not limit recursion for deeply nested JSON documents",
            "description": "orjson.loads in orjson before 3.9.15 does not limit recursion for deeply nested JSON documents.",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-pwr2-4v36-6qpr",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-27454",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-27454"
                },
                {
                    "url": "https://github.com/ijl/orjson/issues/458"
                },
                {
                    "url": "https://github.com/ijl/orjson/commit/b0e4d2c06ce06c6e63981bf0276e4b7c74e5845e"
                },
                {
                    "url": "https://github.com/ijl/orjson/blob/master/CHANGELOG.md#3915"
                },
                {
                    "url": "https://monicz.dev/CVE-2024-27454"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/orjson/PYSEC-2024-40.yaml"
                },
                {
                    "url": "https://github.com/advisories/GHSA-pwr2-4v36-6qpr"
                }
            ],
            "published_at": "2024-02-26T18:30:31Z",
            "updated_at": "2024-02-26T22:06:17Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "orjson"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 3.9.15",
                    "first_patched_version": {
                        "identifier": "3.9.15"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
                "score": 7.5
            },
            "cwes": [
                {
                    "cwe_id": "CWE-674",
                    "name": "Uncontrolled Recursion"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "orjson"
            },
            "severity": "high",
            "vulnerable_version_range": "< 3.9.15",
            "first_patched_version": {
                "identifier": "3.9.15"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/17",
        "html_url": "https://github.com/username/gradio/security/dependabot/17",
        "created_at": "2024-05-24T01:34:13Z",
        "updated_at": "2024-05-24T01:34:13Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 16,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "starlette"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-2jv5-9r88-3w3p",
            "cve_id": "CVE-2024-24762",
            "summary": "python-multipart vulnerable to Content-Type Header ReDoS",
            "description": "### Summary\n\nWhen using form data, `python-multipart` uses a Regular Expression to parse the HTTP `Content-Type` header, including options.\n\nAn attacker could send a custom-made `Content-Type` option that is very difficult for the RegEx to process, consuming CPU resources and stalling indefinitely (minutes or more) while holding the main event loop. This means that process can't handle any more requests.\n\nThis can create a ReDoS (Regular expression Denial of Service): https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS\n\nThis only applies when the app uses form data, parsed with `python-multipart`.\n\n### Details\n\nA regular HTTP `Content-Type` header could look like:\n\n```\nContent-Type: text/html; charset=utf-8\n```\n\n`python-multipart` parses the option with this RegEx: https://github.com/andrew-d/python-multipart/blob/d3d16dae4b061c34fe9d3c9081d9800c49fc1f7a/multipart/multipart.py#L72-L74\n\nA custom option could be made and sent to the server to break it with:\n\n```\nContent-Type: application/x-www-form-urlencoded; !=\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\n```\n\n### PoC\n\nCreate a simple WSGI application, that just parses the `Content-Type`, and run it with `python main.py`:\n\n```Python\n# main.py\nfrom wsgiref.simple_server import make_server\nfrom wsgiref.validate import validator\n\nfrom multipart.multipart import parse_options_header\n\n\ndef simple_app(environ, start_response):\n    _, _ = parse_options_header(environ[\"CONTENT_TYPE\"])\n\n    start_response(\"200 OK\", [(\"Content-type\", \"text/plain\")])\n    return [b\"Ok\"]\n\n\nhttpd = make_server(\"\", 8123, validator(simple_app))\nprint(\"Serving on port 8123...\")\nhttpd.serve_forever()\n```\n\nThen send the attacking request with:\n\n```console\n$ curl -v -X 'POST' -H $'Content-Type: application/x-www-form-urlencoded; !=\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\' --data-binary 'input=1' 'http://localhost:8123/'\n```\n\n### Impact\n\nIt's a ReDoS, (Regular expression Denial of Service), it only applies to those reading form data. This way it also affects other libraries using Starlette, like FastAPI.\n\n### Original Report\n\nThis was originally reported to FastAPI as an email to security@tiangolo.com, sent via https://huntr.com/, the original reporter is Marcello, https://github.com/byt3bl33d3r\n\n<details>\n<summary>Original report to FastAPI</summary>\n\nHey Tiangolo!\n\nMy name's Marcello and I work on the ProtectAI/Huntr Threat Research team, a few months ago we got a report (from @nicecatch2000) of a ReDoS affecting another very popular Python web framework. After some internal research, I found that FastAPI is vulnerable to the same ReDoS under certain conditions (only when it parses Form data not JSON).\n\nHere are the details: I'm using the latest version of FastAPI (0.109.0) and the following code:\n\n```Python\nfrom typing import Annotated\nfrom fastapi.responses import HTMLResponse\nfrom fastapi import FastAPI,Form\nfrom pydantic import BaseModel\n\nclass Item(BaseModel):\n    username: str\n\napp = FastAPI()\n\n@app.get(\"/\", response_class=HTMLResponse)\nasync def index():\n    return HTMLResponse(\"Test\", status_code=200)\n\n@app.post(\"/submit/\")\nasync def submit(username: Annotated[str, Form()]):\n    return {\"username\": username}\n\n@app.post(\"/submit_json/\")\nasync def submit_json(item: Item):\n    return {\"username\": item.username}\n```\n\nI'm running the above with uvicorn with the following command:\n\n```console\nuvicorn server:app\n```\n\nThen run the following cUrl command:\n\n```\ncurl -v -X 'POST' -H $'Content-Type: application/x-www-form-urlencoded; !=\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\' --data-binary 'input=1' 'http://localhost:8000/submit/'\n```\n\nYou'll see the server locks up, is unable to serve anymore requests and one CPU core is pegged to 100%\n\nYou can even start uvicorn with multiple workers with the --workers 4 argument and as long as you send (workers + 1) requests you'll completely DoS the FastApi server.\n\nIf you try submitting Json to the /submit_json endpoint with the malicious Content-Type header you'll see it isn't vulnerable. So this only affects FastAPI when it parses Form data.\n\nCheers\n\n#### Impact\n\nAn attacker is able to cause a DoS on a FastApi server via a malicious Content-Type header if it parses Form data.\n\n#### Occurrences\n\n[params.py L586](https://github.com/tiangolo/fastapi/blob/d74b3b25659b42233a669f032529880de8bd6c2d/fastapi/params.py#L586)\n\n</details>",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-2jv5-9r88-3w3p",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-24762",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/Kludex/python-multipart/security/advisories/GHSA-2jv5-9r88-3w3p"
                },
                {
                    "url": "https://github.com/Kludex/python-multipart/commit/20f0ef6b4e4caf7d69a667c54dff57fe467109a4"
                },
                {
                    "url": "https://github.com/andrew-d/python-multipart/blob/d3d16dae4b061c34fe9d3c9081d9800c49fc1f7a/multipart/multipart.py#L72-L74"
                },
                {
                    "url": "https://github.com/encode/starlette/security/advisories/GHSA-93gm-qmq6-w238"
                },
                {
                    "url": "https://github.com/tiangolo/fastapi/security/advisories/GHSA-qf9m-vfgh-m389"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-24762"
                },
                {
                    "url": "https://github.com/encode/starlette/commit/13e5c26a27f4903924624736abd6131b2da80cc5"
                },
                {
                    "url": "https://github.com/tiangolo/fastapi/commit/9d34ad0ee8a0dfbbcce06f76c2d5d851085024fc"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/fastapi/PYSEC-2024-38.yaml"
                },
                {
                    "url": "https://github.com/tiangolo/fastapi/releases/tag/0.109.1"
                },
                {
                    "url": "https://github.com/advisories/GHSA-2jv5-9r88-3w3p"
                }
            ],
            "published_at": "2024-02-12T17:28:12Z",
            "updated_at": "2024-02-16T23:41:49Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "python-multipart"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "<= 0.0.6",
                    "first_patched_version": {
                        "identifier": "0.0.7"
                    }
                },
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "fastapi"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "<= 0.109.0",
                    "first_patched_version": {
                        "identifier": "0.109.1"
                    }
                },
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "starlette"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "<= 0.36.1",
                    "first_patched_version": {
                        "identifier": "0.36.2"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
                "score": 7.5
            },
            "cwes": [
                {
                    "cwe_id": "CWE-400",
                    "name": "Uncontrolled Resource Consumption"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "starlette"
            },
            "severity": "high",
            "vulnerable_version_range": "<= 0.36.1",
            "first_patched_version": {
                "identifier": "0.36.2"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/16",
        "html_url": "https://github.com/username/gradio/security/dependabot/16",
        "created_at": "2024-05-24T01:34:13Z",
        "updated_at": "2024-05-24T01:34:13Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 15,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "fastapi"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-2jv5-9r88-3w3p",
            "cve_id": "CVE-2024-24762",
            "summary": "python-multipart vulnerable to Content-Type Header ReDoS",
            "description": "### Summary\n\nWhen using form data, `python-multipart` uses a Regular Expression to parse the HTTP `Content-Type` header, including options.\n\nAn attacker could send a custom-made `Content-Type` option that is very difficult for the RegEx to process, consuming CPU resources and stalling indefinitely (minutes or more) while holding the main event loop. This means that process can't handle any more requests.\n\nThis can create a ReDoS (Regular expression Denial of Service): https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS\n\nThis only applies when the app uses form data, parsed with `python-multipart`.\n\n### Details\n\nA regular HTTP `Content-Type` header could look like:\n\n```\nContent-Type: text/html; charset=utf-8\n```\n\n`python-multipart` parses the option with this RegEx: https://github.com/andrew-d/python-multipart/blob/d3d16dae4b061c34fe9d3c9081d9800c49fc1f7a/multipart/multipart.py#L72-L74\n\nA custom option could be made and sent to the server to break it with:\n\n```\nContent-Type: application/x-www-form-urlencoded; !=\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\n```\n\n### PoC\n\nCreate a simple WSGI application, that just parses the `Content-Type`, and run it with `python main.py`:\n\n```Python\n# main.py\nfrom wsgiref.simple_server import make_server\nfrom wsgiref.validate import validator\n\nfrom multipart.multipart import parse_options_header\n\n\ndef simple_app(environ, start_response):\n    _, _ = parse_options_header(environ[\"CONTENT_TYPE\"])\n\n    start_response(\"200 OK\", [(\"Content-type\", \"text/plain\")])\n    return [b\"Ok\"]\n\n\nhttpd = make_server(\"\", 8123, validator(simple_app))\nprint(\"Serving on port 8123...\")\nhttpd.serve_forever()\n```\n\nThen send the attacking request with:\n\n```console\n$ curl -v -X 'POST' -H $'Content-Type: application/x-www-form-urlencoded; !=\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\' --data-binary 'input=1' 'http://localhost:8123/'\n```\n\n### Impact\n\nIt's a ReDoS, (Regular expression Denial of Service), it only applies to those reading form data. This way it also affects other libraries using Starlette, like FastAPI.\n\n### Original Report\n\nThis was originally reported to FastAPI as an email to security@tiangolo.com, sent via https://huntr.com/, the original reporter is Marcello, https://github.com/byt3bl33d3r\n\n<details>\n<summary>Original report to FastAPI</summary>\n\nHey Tiangolo!\n\nMy name's Marcello and I work on the ProtectAI/Huntr Threat Research team, a few months ago we got a report (from @nicecatch2000) of a ReDoS affecting another very popular Python web framework. After some internal research, I found that FastAPI is vulnerable to the same ReDoS under certain conditions (only when it parses Form data not JSON).\n\nHere are the details: I'm using the latest version of FastAPI (0.109.0) and the following code:\n\n```Python\nfrom typing import Annotated\nfrom fastapi.responses import HTMLResponse\nfrom fastapi import FastAPI,Form\nfrom pydantic import BaseModel\n\nclass Item(BaseModel):\n    username: str\n\napp = FastAPI()\n\n@app.get(\"/\", response_class=HTMLResponse)\nasync def index():\n    return HTMLResponse(\"Test\", status_code=200)\n\n@app.post(\"/submit/\")\nasync def submit(username: Annotated[str, Form()]):\n    return {\"username\": username}\n\n@app.post(\"/submit_json/\")\nasync def submit_json(item: Item):\n    return {\"username\": item.username}\n```\n\nI'm running the above with uvicorn with the following command:\n\n```console\nuvicorn server:app\n```\n\nThen run the following cUrl command:\n\n```\ncurl -v -X 'POST' -H $'Content-Type: application/x-www-form-urlencoded; !=\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\' --data-binary 'input=1' 'http://localhost:8000/submit/'\n```\n\nYou'll see the server locks up, is unable to serve anymore requests and one CPU core is pegged to 100%\n\nYou can even start uvicorn with multiple workers with the --workers 4 argument and as long as you send (workers + 1) requests you'll completely DoS the FastApi server.\n\nIf you try submitting Json to the /submit_json endpoint with the malicious Content-Type header you'll see it isn't vulnerable. So this only affects FastAPI when it parses Form data.\n\nCheers\n\n#### Impact\n\nAn attacker is able to cause a DoS on a FastApi server via a malicious Content-Type header if it parses Form data.\n\n#### Occurrences\n\n[params.py L586](https://github.com/tiangolo/fastapi/blob/d74b3b25659b42233a669f032529880de8bd6c2d/fastapi/params.py#L586)\n\n</details>",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-2jv5-9r88-3w3p",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-24762",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/Kludex/python-multipart/security/advisories/GHSA-2jv5-9r88-3w3p"
                },
                {
                    "url": "https://github.com/Kludex/python-multipart/commit/20f0ef6b4e4caf7d69a667c54dff57fe467109a4"
                },
                {
                    "url": "https://github.com/andrew-d/python-multipart/blob/d3d16dae4b061c34fe9d3c9081d9800c49fc1f7a/multipart/multipart.py#L72-L74"
                },
                {
                    "url": "https://github.com/encode/starlette/security/advisories/GHSA-93gm-qmq6-w238"
                },
                {
                    "url": "https://github.com/tiangolo/fastapi/security/advisories/GHSA-qf9m-vfgh-m389"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-24762"
                },
                {
                    "url": "https://github.com/encode/starlette/commit/13e5c26a27f4903924624736abd6131b2da80cc5"
                },
                {
                    "url": "https://github.com/tiangolo/fastapi/commit/9d34ad0ee8a0dfbbcce06f76c2d5d851085024fc"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/fastapi/PYSEC-2024-38.yaml"
                },
                {
                    "url": "https://github.com/tiangolo/fastapi/releases/tag/0.109.1"
                },
                {
                    "url": "https://github.com/advisories/GHSA-2jv5-9r88-3w3p"
                }
            ],
            "published_at": "2024-02-12T17:28:12Z",
            "updated_at": "2024-02-16T23:41:49Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "python-multipart"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "<= 0.0.6",
                    "first_patched_version": {
                        "identifier": "0.0.7"
                    }
                },
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "fastapi"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "<= 0.109.0",
                    "first_patched_version": {
                        "identifier": "0.109.1"
                    }
                },
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "starlette"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "<= 0.36.1",
                    "first_patched_version": {
                        "identifier": "0.36.2"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
                "score": 7.5
            },
            "cwes": [
                {
                    "cwe_id": "CWE-400",
                    "name": "Uncontrolled Resource Consumption"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "fastapi"
            },
            "severity": "high",
            "vulnerable_version_range": "<= 0.109.0",
            "first_patched_version": {
                "identifier": "0.109.1"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/15",
        "html_url": "https://github.com/username/gradio/security/dependabot/15",
        "created_at": "2024-05-24T01:34:13Z",
        "updated_at": "2024-05-24T01:34:13Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 14,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "Pillow"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-3f63-hfp8-52jq",
            "cve_id": "CVE-2023-50447",
            "summary": "Arbitrary Code Execution in Pillow",
            "description": "Pillow through 10.1.0 allows PIL.ImageMath.eval Arbitrary Code Execution via the environment parameter, a different vulnerability than CVE-2022-22817 (which was about the expression parameter).",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-3f63-hfp8-52jq",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2023-50447",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-50447"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/commit/45c726fd4daa63236a8f3653530f297dc87b160a"
                },
                {
                    "url": "https://pillow.readthedocs.io/en/stable/releasenotes/10.2.0.html#security"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2024/01/20/1"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/releases"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2024/01/msg00019.html"
                },
                {
                    "url": "https://devhub.checkmarx.com/cve-details/CVE-2023-50447"
                },
                {
                    "url": "https://duartecsantos.github.io/2023-01-02-CVE-2023-50447"
                },
                {
                    "url": "https://duartecsantos.github.io/2024-01-02-CVE-2023-50447"
                },
                {
                    "url": "https://github.com/advisories/GHSA-3f63-hfp8-52jq"
                }
            ],
            "published_at": "2024-01-19T21:30:35Z",
            "updated_at": "2024-03-27T23:53:02Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "Pillow"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 10.2.0",
                    "first_patched_version": {
                        "identifier": "10.2.0"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:H/I:H/A:H",
                "score": 8.1
            },
            "cwes": [
                {
                    "cwe_id": "CWE-94",
                    "name": "Improper Control of Generation of Code ('Code Injection')"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "Pillow"
            },
            "severity": "high",
            "vulnerable_version_range": "< 10.2.0",
            "first_patched_version": {
                "identifier": "10.2.0"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/14",
        "html_url": "https://github.com/username/gradio/security/dependabot/14",
        "created_at": "2024-05-24T01:34:13Z",
        "updated_at": "2024-05-24T01:34:13Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 13,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "jinja2"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-h5c8-rqwp-cp95",
            "cve_id": "CVE-2024-22195",
            "summary": "Jinja vulnerable to HTML attribute injection when passing user input as keys to xmlattr filter",
            "description": "The `xmlattr` filter in affected versions of Jinja accepts keys containing spaces. XML/HTML attributes cannot contain spaces, as each would then be interpreted as a separate attribute. If an application accepts keys (as opposed to only values) as user input, and renders these in pages that other users see as well, an attacker could use this to inject other attributes and perform XSS. Note that accepting keys as user input is not common or a particularly intended use case of the `xmlattr` filter, and an application doing so should already be verifying what keys are provided regardless of this fix.",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-h5c8-rqwp-cp95",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2024-22195",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/pallets/jinja/security/advisories/GHSA-h5c8-rqwp-cp95"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2024-22195"
                },
                {
                    "url": "https://github.com/pallets/jinja/commit/716795349a41d4983a9a4771f7d883c96ea17be7"
                },
                {
                    "url": "https://github.com/pallets/jinja/releases/tag/3.1.3"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/5XCWZD464AJJJUBOO7CMPXQ4ROBC6JX2/"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2024/01/msg00010.html"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/DELCVUUYX75I5K4Q5WMJG4MUZJA6VAIP/"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/O7YWRBX6JQCWC2XXCTZ55C7DPMGICCN3/"
                },
                {
                    "url": "https://github.com/advisories/GHSA-h5c8-rqwp-cp95"
                }
            ],
            "published_at": "2024-01-11T15:20:48Z",
            "updated_at": "2024-01-27T03:30:20Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "jinja2"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": "< 3.1.3",
                    "first_patched_version": {
                        "identifier": "3.1.3"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:U/C:L/I:L/A:N",
                "score": 5.4
            },
            "cwes": [
                {
                    "cwe_id": "CWE-79",
                    "name": "Improper Neutralization of Input During Web Page Generation ('Cross-site Scripting')"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "jinja2"
            },
            "severity": "medium",
            "vulnerable_version_range": "< 3.1.3",
            "first_patched_version": {
                "identifier": "3.1.3"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/13",
        "html_url": "https://github.com/username/gradio/security/dependabot/13",
        "created_at": "2024-05-24T01:34:13Z",
        "updated_at": "2024-05-24T01:34:13Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 12,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "pillow"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-8ghj-p4vj-mr35",
            "cve_id": "CVE-2023-44271",
            "summary": "Pillow Denial of Service vulnerability",
            "description": "An issue was discovered in Pillow before 10.0.0. It is a Denial of Service that uncontrollably allocates memory to process a given task, potentially causing a service to crash by having it run out of memory. This occurs for truetype in ImageFont when textlength in an ImageDraw instance operates on a long text argument.",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-8ghj-p4vj-mr35",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2023-44271",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-44271"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/pull/7244"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/commit/1fe1bb49c452b0318cad12ea9d97c3bef188e9a7"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/pillow/PYSEC-2023-227.yaml"
                },
                {
                    "url": "https://devhub.checkmarx.com/cve-details/CVE-2023-44271"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2024/03/msg00021.html"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce%40lists.fedoraproject.org/message/N2JOEDUJDQLCUII2LQYZYSM7RJL2I3P4"
                },
                {
                    "url": "https://github.com/advisories/GHSA-8ghj-p4vj-mr35"
                }
            ],
            "published_at": "2023-11-03T06:36:30Z",
            "updated_at": "2024-03-22T15:25:43Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "pillow"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 0, < 10.0.0",
                    "first_patched_version": {
                        "identifier": "10.0.0"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
                "score": 7.5
            },
            "cwes": [
                {
                    "cwe_id": "CWE-400",
                    "name": "Uncontrolled Resource Consumption"
                },
                {
                    "cwe_id": "CWE-770",
                    "name": "Allocation of Resources Without Limits or Throttling"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "pillow"
            },
            "severity": "high",
            "vulnerable_version_range": ">= 0, < 10.0.0",
            "first_patched_version": {
                "identifier": "10.0.0"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/12",
        "html_url": "https://github.com/username/gradio/security/dependabot/12",
        "created_at": "2024-05-24T01:34:13Z",
        "updated_at": "2024-05-24T01:34:13Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 11,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "urllib3"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-g4mx-q9vg-27p4",
            "cve_id": "CVE-2023-45803",
            "summary": "urllib3's request body not stripped after redirect from 303 status changes request method to GET",
            "description": "urllib3 previously wouldn't remove the HTTP request body when an HTTP redirect response using status 303 \"See Other\" after the request had its method changed from one that could accept a request body (like `POST`) to `GET` as is required by HTTP RFCs. Although the behavior of removing the request body is not specified in the section for redirects, it can be inferred by piecing together information from different sections and we have observed the behavior in other major HTTP client implementations like curl and web browsers.\n\nFrom [RFC 9110 Section 9.3.1](https://www.rfc-editor.org/rfc/rfc9110.html#name-get):\n\n> A client SHOULD NOT generate content in a GET request unless it is made directly to an origin server that has previously indicated, in or out of band, that such a request has a purpose and will be adequately supported.\n\n## Affected usages\n\nBecause the vulnerability requires a previously trusted service to become compromised in order to have an impact on confidentiality we believe the exploitability of this vulnerability is low. Additionally, many users aren't putting sensitive data in HTTP request bodies, if this is the case then this vulnerability isn't exploitable.\n\nBoth of the following conditions must be true to be affected by this vulnerability:\n\n* If you're using urllib3 and submitting sensitive information in the HTTP request body (such as form data or JSON)\n* The origin service is compromised and starts redirecting using 303 to a malicious peer or the redirected-to service becomes compromised.\n\n## Remediation\n\nYou can remediate this vulnerability with any of the following steps:\n\n* Upgrade to a patched version of urllib3 (v1.26.18 or v2.0.7)\n* Disable redirects for services that you aren't expecting to respond with redirects with `redirects=False`.\n* Disable automatic redirects with `redirects=False` and handle 303 redirects manually by stripping the HTTP request body.\n",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-g4mx-q9vg-27p4",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2023-45803",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/urllib3/urllib3/security/advisories/GHSA-g4mx-q9vg-27p4"
                },
                {
                    "url": "https://github.com/urllib3/urllib3/commit/4e50fbc5db74e32cabd5ccc1ab81fc103adfe0b3"
                },
                {
                    "url": "https://github.com/urllib3/urllib3/commit/b594c5ceaca38e1ac215f916538fb128e3526a36"
                },
                {
                    "url": "https://github.com/urllib3/urllib3/releases/tag/1.26.18"
                },
                {
                    "url": "https://github.com/urllib3/urllib3/releases/tag/2.0.7"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-45803"
                },
                {
                    "url": "https://github.com/urllib3/urllib3/commit/4e98d57809dacab1cbe625fddeec1a290c478ea9"
                },
                {
                    "url": "https://www.rfc-editor.org/rfc/rfc9110.html#name-get"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/PPDPLM6UUMN55ESPQWJFLLIZY4ZKCNRX/"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/urllib3/PYSEC-2023-212.yaml"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/4R2Y5XK3WALSR3FNAGN7JBYV2B343ZKB/"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/5F5CUBAN5XMEBVBZPHFITBLMJV5FIJJ5/"
                },
                {
                    "url": "https://github.com/advisories/GHSA-g4mx-q9vg-27p4"
                }
            ],
            "published_at": "2023-10-17T20:15:25Z",
            "updated_at": "2023-11-06T16:24:46Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "urllib3"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": ">= 2.0.0, < 2.0.7",
                    "first_patched_version": {
                        "identifier": "2.0.7"
                    }
                },
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "urllib3"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": ">= 0, < 1.26.18",
                    "first_patched_version": {
                        "identifier": "1.26.18"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:A/AC:H/PR:H/UI:N/S:U/C:H/I:N/A:N",
                "score": 4.2
            },
            "cwes": [
                {
                    "cwe_id": "CWE-200",
                    "name": "Exposure of Sensitive Information to an Unauthorized Actor"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "urllib3"
            },
            "severity": "medium",
            "vulnerable_version_range": ">= 0, < 1.26.18",
            "first_patched_version": {
                "identifier": "1.26.18"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/11",
        "html_url": "https://github.com/username/gradio/security/dependabot/11",
        "created_at": "2024-05-24T01:34:13Z",
        "updated_at": "2024-05-24T01:34:13Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 10,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "pillow"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-56pw-mpj4-fxww",
            "cve_id": null,
            "summary": "Bundled libwebp in Pillow vulnerable",
            "description": "Pillow versions before v10.0.1 bundled libwebp binaries in wheels that are vulnerable to CVE-2023-5129 (previously CVE-2023-4863). Pillow v10.0.1 upgrades the bundled libwebp binary to v1.3.2.",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-56pw-mpj4-fxww",
                    "type": "GHSA"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-4863"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-5129"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/pillow/PYSEC-2023-175.yaml"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/blob/main/CHANGES.rst#1001-2023-09-15"
                },
                {
                    "url": "https://github.com/advisories/GHSA-56pw-mpj4-fxww"
                }
            ],
            "published_at": "2023-10-05T00:06:58Z",
            "updated_at": "2023-10-05T00:06:58Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "pillow"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 10.0.1",
                    "first_patched_version": {
                        "identifier": "10.0.1"
                    }
                }
            ],
            "cvss": {
                "vector_string": null,
                "score": 0.0
            },
            "cwes": []
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "pillow"
            },
            "severity": "high",
            "vulnerable_version_range": "< 10.0.1",
            "first_patched_version": {
                "identifier": "10.0.1"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/10",
        "html_url": "https://github.com/username/gradio/security/dependabot/10",
        "created_at": "2024-05-24T01:34:13Z",
        "updated_at": "2024-05-24T01:34:13Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 9,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "Pillow"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-j7hp-h8jx-5ppr",
            "cve_id": "CVE-2023-4863",
            "summary": "libwebp: OOB write in BuildHuffmanTable",
            "description": "Heap buffer overflow in libwebp allow a remote attacker to perform an out of bounds memory write via a crafted HTML page. ",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-j7hp-h8jx-5ppr",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2023-4863",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-4863"
                },
                {
                    "url": "https://chromereleases.googleblog.com/2023/09/stable-channel-update-for-desktop_11.html"
                },
                {
                    "url": "https://crbug.com/1479274"
                },
                {
                    "url": "https://github.com/qnighy/libwebp-sys2-rs/pull/21"
                },
                {
                    "url": "https://github.com/qnighy/libwebp-sys2-rs/commit/4560c473a76ec8bd8c650f19ddf9d7a44f719f8b"
                },
                {
                    "url": "https://github.com/webmproject/libwebp/commit/902bc9190331343b2017211debcec8d2ab87e17a"
                },
                {
                    "url": "https://bugzilla.suse.com/show_bug.cgi?id=1215231"
                },
                {
                    "url": "https://en.bandisoft.com/honeyview/history/"
                },
                {
                    "url": "https://msrc.microsoft.com/update-guide/vulnerability/CVE-2023-4863"
                },
                {
                    "url": "https://news.ycombinator.com/item?id=37478403"
                },
                {
                    "url": "https://rustsec.org/advisories/RUSTSEC-2023-0060.html"
                },
                {
                    "url": "https://security-tracker.debian.org/tracker/CVE-2023-4863"
                },
                {
                    "url": "https://stackdiary.com/critical-vulnerability-in-webp-codec-cve-2023-4863/"
                },
                {
                    "url": "https://www.bleepingcomputer.com/news/google/google-fixes-another-chrome-zero-day-bug-exploited-in-attacks/"
                },
                {
                    "url": "https://www.mozilla.org/en-US/security/advisories/mfsa2023-40/"
                },
                {
                    "url": "https://github.com/electron/electron/pull/39823"
                },
                {
                    "url": "https://github.com/electron/electron/pull/39825"
                },
                {
                    "url": "https://github.com/electron/electron/pull/39826"
                },
                {
                    "url": "https://github.com/electron/electron/pull/39827"
                },
                {
                    "url": "https://github.com/electron/electron/pull/39828"
                },
                {
                    "url": "https://adamcaudill.com/2023/09/14/whose-cve-is-it-anyway/"
                },
                {
                    "url": "https://github.com/webmproject/libwebp/releases/tag/v1.3.2"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2023/09/msg00015.html"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2023/09/msg00016.html"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2023/09/msg00017.html"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/6T655QF7CQ3DYAMPFV7IECQYGDEUIVVT/"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/FYYKLG6CRGEDTNRBSU26EEWAO6D6U645/"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/OZDGWWMJREPAGKWCJKSCM4WYLANSKIFX/"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/PYZV7TMKF4QHZ54SFJX54BDN52VHGGCX/"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/WHOLML7N2G5KCAZXFWC5IDFFHSQS5SDB/"
                },
                {
                    "url": "https://rustsec.org/advisories/RUSTSEC-2023-0061.html"
                },
                {
                    "url": "https://security.gentoo.org/glsa/202309-05"
                },
                {
                    "url": "https://www.debian.org/security/2023/dsa-5496"
                },
                {
                    "url": "https://www.debian.org/security/2023/dsa-5497"
                },
                {
                    "url": "https://www.debian.org/security/2023/dsa-5498"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/KUQ7CTX3W372X3UY56VVNAHCH6H2F4X3/"
                },
                {
                    "url": "https://blog.isosceles.com/the-webp-0day/"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/21/4"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/22/1"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/22/3"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/22/4"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/22/5"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/22/6"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/22/7"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/22/8"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/26/1"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/26/7"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/28/1"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/28/2"
                },
                {
                    "url": "http://www.openwall.com/lists/oss-security/2023/09/28/4"
                },
                {
                    "url": "https://github.com/jaredforth/webp/pull/30"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/pull/7395"
                },
                {
                    "url": "https://github.com/jaredforth/webp/commit/9d4c56e63abecc777df71c702503c3eaabd7dcbc"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/WTRUIS3564P7ZLM2S2IH4Y4KZ327LI4I/"
                },
                {
                    "url": "https://pillow.readthedocs.io/en/stable/releasenotes/10.0.1.html#security"
                },
                {
                    "url": "https://security.netapp.com/advisory/ntap-20230929-0011/"
                },
                {
                    "url": "https://github.com/ImageMagick/ImageMagick/discussions/6664"
                },
                {
                    "url": "https://github.com/dlemstra/Magick.NET/releases/tag/13.3.0"
                },
                {
                    "url": "https://security.gentoo.org/glsa/202401-10"
                },
                {
                    "url": "https://sethmlarson.dev/security-developer-in-residence-weekly-report-16"
                },
                {
                    "url": "https://www.bentley.com/advisories/be-2023-0001/"
                },
                {
                    "url": "https://github.com/advisories/GHSA-j7hp-h8jx-5ppr"
                }
            ],
            "published_at": "2023-09-12T15:30:20Z",
            "updated_at": "2024-01-08T15:38:39Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "rust",
                        "name": "libwebp-sys2"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 0.1.8",
                    "first_patched_version": {
                        "identifier": "0.1.8"
                    }
                },
                {
                    "package": {
                        "ecosystem": "rust",
                        "name": "libwebp-sys"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 0.9.3",
                    "first_patched_version": {
                        "identifier": "0.9.3"
                    }
                },
                {
                    "package": {
                        "ecosystem": "npm",
                        "name": "electron"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 22.0.0, < 22.3.24",
                    "first_patched_version": {
                        "identifier": "22.3.24"
                    }
                },
                {
                    "package": {
                        "ecosystem": "npm",
                        "name": "electron"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 24.0.0, < 24.8.3",
                    "first_patched_version": {
                        "identifier": "24.8.3"
                    }
                },
                {
                    "package": {
                        "ecosystem": "npm",
                        "name": "electron"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 25.0.0, < 25.8.1",
                    "first_patched_version": {
                        "identifier": "25.8.1"
                    }
                },
                {
                    "package": {
                        "ecosystem": "npm",
                        "name": "electron"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 26.0.0, < 26.2.1",
                    "first_patched_version": {
                        "identifier": "26.2.1"
                    }
                },
                {
                    "package": {
                        "ecosystem": "npm",
                        "name": "electron"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 27.0.0-beta.1, < 27.0.0-beta.2",
                    "first_patched_version": {
                        "identifier": "27.0.0-beta.2"
                    }
                },
                {
                    "package": {
                        "ecosystem": "nuget",
                        "name": "SkiaSharp"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 2.0.0, < 2.88.6",
                    "first_patched_version": {
                        "identifier": "2.88.6"
                    }
                },
                {
                    "package": {
                        "ecosystem": "go",
                        "name": "github.com/chai2010/webp"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 1.0.0",
                    "first_patched_version": null
                },
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "Pillow"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 10.0.1",
                    "first_patched_version": {
                        "identifier": "10.0.1"
                    }
                },
                {
                    "package": {
                        "ecosystem": "rust",
                        "name": "webp"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 0.2.6",
                    "first_patched_version": {
                        "identifier": "0.2.6"
                    }
                },
                {
                    "package": {
                        "ecosystem": "nuget",
                        "name": "magick.net-q16-anycpu"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 13.3.0",
                    "first_patched_version": {
                        "identifier": "13.3.0"
                    }
                },
                {
                    "package": {
                        "ecosystem": "nuget",
                        "name": "magick.net-q16-hdri-anycpu"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 13.3.0",
                    "first_patched_version": {
                        "identifier": "13.3.0"
                    }
                },
                {
                    "package": {
                        "ecosystem": "nuget",
                        "name": "magick.net-q16-x64"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 13.3.0",
                    "first_patched_version": {
                        "identifier": "13.3.0"
                    }
                },
                {
                    "package": {
                        "ecosystem": "nuget",
                        "name": "magick.net-q8-anycpu"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 13.3.0",
                    "first_patched_version": {
                        "identifier": "13.3.0"
                    }
                },
                {
                    "package": {
                        "ecosystem": "nuget",
                        "name": "magick.net-q8-openmp-x64"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 13.3.0",
                    "first_patched_version": {
                        "identifier": "13.3.0"
                    }
                },
                {
                    "package": {
                        "ecosystem": "nuget",
                        "name": "magick.net-q8-x64"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "< 13.3.0",
                    "first_patched_version": {
                        "identifier": "13.3.0"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:U/C:H/I:H/A:H",
                "score": 8.8
            },
            "cwes": [
                {
                    "cwe_id": "CWE-787",
                    "name": "Out-of-bounds Write"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "Pillow"
            },
            "severity": "high",
            "vulnerable_version_range": "< 10.0.1",
            "first_patched_version": {
                "identifier": "10.0.1"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/9",
        "html_url": "https://github.com/username/gradio/security/dependabot/9",
        "created_at": "2024-05-24T01:34:12Z",
        "updated_at": "2024-05-24T01:34:12Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 8,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "urllib3"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-v845-jxx5-vc9f",
            "cve_id": "CVE-2023-43804",
            "summary": "`Cookie` HTTP header isn't stripped on cross-origin redirects",
            "description": "urllib3 doesn't treat the `Cookie` HTTP header special or provide any helpers for managing cookies over HTTP, that is the responsibility of the user. However, it is possible for a user to specify a `Cookie` header and unknowingly leak information via HTTP redirects to a different origin if that user doesn't disable redirects explicitly.\n\nUsers **must** handle redirects themselves instead of relying on urllib3's automatic redirects to achieve safe processing of the `Cookie` header, thus we decided to strip the header by default in order to further protect users who aren't using the correct approach.\n\n## Affected usages\n\nWe believe the number of usages affected by this advisory is low. It requires all of the following to be true to be exploited:\n\n* Using an affected version of urllib3 (patched in v1.26.17 and v2.0.6)\n* Using the `Cookie` header on requests, which is mostly typical for impersonating a browser.\n* Not disabling HTTP redirects\n* Either not using HTTPS or for the origin server to redirect to a malicious origin.\n\n## Remediation\n\n* Upgrading to at least urllib3 v1.26.17 or v2.0.6\n* Disabling HTTP redirects using `redirects=False` when sending requests.\n* Not using the `Cookie` header.",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-v845-jxx5-vc9f",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2023-43804",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/urllib3/urllib3/security/advisories/GHSA-v845-jxx5-vc9f"
                },
                {
                    "url": "https://github.com/urllib3/urllib3/commit/01220354d389cd05474713f8c982d05c9b17aafb"
                },
                {
                    "url": "https://github.com/urllib3/urllib3/commit/644124ecd0b6e417c527191f866daa05a5a2056d"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-43804"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2023/10/msg00012.html"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/urllib3/PYSEC-2023-192.yaml"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/I3PR7C6RJ6JUBQKIJ644DMIJSUP36VDY/"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/NDAGZXYJ7H2G3SB47M453VQVNAWKAEJJ/"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/5F5CUBAN5XMEBVBZPHFITBLMJV5FIJJ5/"
                },
                {
                    "url": "https://github.com/advisories/GHSA-v845-jxx5-vc9f"
                }
            ],
            "published_at": "2023-10-02T23:27:05Z",
            "updated_at": "2023-11-04T00:30:22Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "urllib3"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": ">= 2.0.0, < 2.0.6",
                    "first_patched_version": {
                        "identifier": "2.0.6"
                    }
                },
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "urllib3"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": "< 1.26.17",
                    "first_patched_version": {
                        "identifier": "1.26.17"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:H/PR:H/UI:N/S:U/C:H/I:H/A:N",
                "score": 5.9
            },
            "cwes": [
                {
                    "cwe_id": "CWE-200",
                    "name": "Exposure of Sensitive Information to an Unauthorized Actor"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "urllib3"
            },
            "severity": "medium",
            "vulnerable_version_range": "< 1.26.17",
            "first_patched_version": {
                "identifier": "1.26.17"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/8",
        "html_url": "https://github.com/username/gradio/security/dependabot/8",
        "created_at": "2024-05-24T01:34:12Z",
        "updated_at": "2024-05-24T01:34:12Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 7,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "certifi"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-xqr8-7jwr-rhp7",
            "cve_id": "CVE-2023-37920",
            "summary": "Removal of e-Tugra root certificate",
            "description": "Certifi 2023.07.22 removes root certificates from \"e-Tugra\" from the root store. These are in the process of being removed from Mozilla's trust store.\n\n e-Tugra's root certificates are being removed pursuant to an investigation prompted by reporting of security issues in their systems. Conclusions of Mozilla's investigation can be found [here](https://groups.google.com/a/mozilla.org/g/dev-security-policy/c/C-HrP1SEq1A).",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-xqr8-7jwr-rhp7",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2023-37920",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/certifi/python-certifi/security/advisories/GHSA-xqr8-7jwr-rhp7"
                },
                {
                    "url": "https://github.com/certifi/python-certifi/commit/8fb96ed81f71e7097ed11bc4d9b19afd7ea5c909"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-37920"
                },
                {
                    "url": "https://groups.google.com/a/mozilla.org/g/dev-security-policy/c/C-HrP1SEq1A"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/certifi/PYSEC-2023-135.yaml"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/5EX6NG7WUFNUKGFHLM35KHHU3GAKXRTG/"
                },
                {
                    "url": "https://github.com/advisories/GHSA-xqr8-7jwr-rhp7"
                }
            ],
            "published_at": "2023-07-25T14:43:53Z",
            "updated_at": "2023-11-09T05:05:47Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "certifi"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 2015.4.28, < 2023.7.22",
                    "first_patched_version": {
                        "identifier": "2023.7.22"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:H/A:N",
                "score": 7.5
            },
            "cwes": [
                {
                    "cwe_id": "CWE-345",
                    "name": "Insufficient Verification of Data Authenticity"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "certifi"
            },
            "severity": "high",
            "vulnerable_version_range": ">= 2015.4.28, < 2023.7.22",
            "first_patched_version": {
                "identifier": "2023.7.22"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/7",
        "html_url": "https://github.com/username/gradio/security/dependabot/7",
        "created_at": "2024-05-24T01:34:12Z",
        "updated_at": "2024-05-24T01:34:12Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 6,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "requests"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-j8r2-6x86-q33q",
            "cve_id": "CVE-2023-32681",
            "summary": "Unintended leak of Proxy-Authorization header in requests",
            "description": "### Impact\n\nSince Requests v2.3.0, Requests has been vulnerable to potentially leaking `Proxy-Authorization` headers to destination servers, specifically during redirects to an HTTPS origin. This is a product of how `rebuild_proxies` is used to recompute and [reattach the `Proxy-Authorization` header](https://github.com/psf/requests/blob/f2629e9e3c7ce3c3c8c025bcd8db551101cbc773/requests/sessions.py#L319-L328) to requests when redirected. Note this behavior has _only_ been observed to affect proxied requests when credentials are supplied in the URL user information component (e.g. `https://username:password@proxy:8080`).\n\n**Current vulnerable behavior(s):**\n\n1. HTTP → HTTPS: **leak**\n2. HTTPS → HTTP: **no leak**\n3. HTTPS → HTTPS: **leak**\n4. HTTP → HTTP: **no leak**\n\nFor HTTP connections sent through the proxy, the proxy will identify the header in the request itself and remove it prior to forwarding to the destination server. However when sent over HTTPS, the `Proxy-Authorization` header must be sent in the CONNECT request as the proxy has no visibility into further tunneled requests. This results in Requests forwarding the header to the destination server unintentionally, allowing a malicious actor to potentially exfiltrate those credentials.\n\nThe reason this currently works for HTTPS connections in Requests is the `Proxy-Authorization` header is also handled by urllib3 with our usage of the ProxyManager in adapters.py with [`proxy_manager_for`](https://github.com/psf/requests/blob/f2629e9e3c7ce3c3c8c025bcd8db551101cbc773/requests/adapters.py#L199-L235). This will compute the required proxy headers in `proxy_headers` and pass them to the Proxy Manager, avoiding attaching them directly to the Request object. This will be our preferred option going forward for default usage.\n\n### Patches\nStarting in Requests v2.31.0, Requests will no longer attach this header to redirects with an HTTPS destination. This should have no negative impacts on the default behavior of the library as the proxy credentials are already properly being handled by urllib3's ProxyManager.\n\nFor users with custom adapters, this _may_ be potentially breaking if you were already working around this behavior. The previous functionality of `rebuild_proxies` doesn't make sense in any case, so we would encourage any users impacted to migrate any handling of Proxy-Authorization directly into their custom adapter.\n\n### Workarounds\nFor users who are not able to update Requests immediately, there is one potential workaround.\n\nYou may disable redirects by setting `allow_redirects` to `False` on all calls through Requests top-level APIs. Note that if you're currently relying on redirect behaviors, you will need to capture the 3xx response codes and ensure a new request is made to the redirect destination.\n```\nimport requests\nr = requests.get('http://github.com/', allow_redirects=False)\n```\n\n### Credits\n\nThis vulnerability was discovered and disclosed by the following individuals.\n\nDennis Brinkrolf, Haxolot (https://haxolot.com/)\nTobias Funke, (tobiasfunke93@gmail.com)",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-j8r2-6x86-q33q",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2023-32681",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/psf/requests/security/advisories/GHSA-j8r2-6x86-q33q"
                },
                {
                    "url": "https://github.com/psf/requests/commit/74ea7cf7a6a27a4eeb2ae24e162bcc942a6706d5"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-32681"
                },
                {
                    "url": "https://github.com/psf/requests/releases/tag/v2.31.0"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/requests/PYSEC-2023-74.yaml"
                },
                {
                    "url": "https://lists.debian.org/debian-lts-announce/2023/06/msg00018.html"
                },
                {
                    "url": "https://security.gentoo.org/glsa/202309-08"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/AW7HNFGYP44RT3DUDQXG2QT3OEV2PJ7Y"
                },
                {
                    "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/KOYASTZDGQG2BWLSNBPL3TQRL2G7QYNZ"
                },
                {
                    "url": "https://github.com/advisories/GHSA-j8r2-6x86-q33q"
                }
            ],
            "published_at": "2023-05-22T20:36:32Z",
            "updated_at": "2024-03-27T20:09:55Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "requests"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": ">= 2.3.0, < 2.31.0",
                    "first_patched_version": {
                        "identifier": "2.31.0"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:H/PR:N/UI:R/S:C/C:H/I:N/A:N",
                "score": 6.1
            },
            "cwes": [
                {
                    "cwe_id": "CWE-200",
                    "name": "Exposure of Sensitive Information to an Unauthorized Actor"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "requests"
            },
            "severity": "medium",
            "vulnerable_version_range": ">= 2.3.0, < 2.31.0",
            "first_patched_version": {
                "identifier": "2.31.0"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/6",
        "html_url": "https://github.com/username/gradio/security/dependabot/6",
        "created_at": "2024-05-24T01:34:12Z",
        "updated_at": "2024-05-24T01:34:12Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 5,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "IPython"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-29gw-9793-fvw7",
            "cve_id": "CVE-2023-24816",
            "summary": "IPython vulnerable to command injection via set_term_title",
            "description": "IPython provides an interactive Python shell and Jupyter kernel to use Python interactively. Versions prior to 8.10.0 are vulnerable to command injection in the `set_term_title` [function](https://github.com/ipython/ipython/blob/3f0bf05f072a91b2a3042d23ce250e5e906183fd/IPython/utils/terminal.py#L103-L117) under specific conditions. This has been patched in version 8.10.0.\n\n### Impact \nUsers are only vulnerable when calling this function in Windows in a Python environment where [ctypes](https://docs.python.org/3/library/ctypes.html) is not available. The dependency on ctypes in `IPython.utils._process_win32` prevents the vulnerable code from ever being reached (making it effectively dead code). However, as a library that could be used by another tool, `set_term_title` could introduce a vulnerability for dependencies. Currently `set_term_title` is only called with (semi-)trusted input that contain the current working directory of the current IPython session. If an attacker can control directory names, and manage to get a user to `cd` into this directory, then the attacker can execute arbitrary commands contained in the folder names.",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-29gw-9793-fvw7",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2023-24816",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/ipython/ipython/security/advisories/GHSA-29gw-9793-fvw7"
                },
                {
                    "url": "https://github.com/ipython/ipython/commit/385d69325319a5972ee9b5983638e3617f21cb1f"
                },
                {
                    "url": "https://github.com/ipython/ipython/commit/991849c247fc208628879e7ca2923b3c218a5a75"
                },
                {
                    "url": "https://github.com/Carreau/ipython/blob/7557ade0ed927475d5ab5b573d0ea4febfb22683/docs/source/whatsnew/version8.rst#ipython-810"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-24816"
                },
                {
                    "url": "https://github.com/ipython/ipython/blob/3f0bf05f072a91b2a3042d23ce250e5e906183fd/IPython/utils/terminal.py#L103-L117"
                },
                {
                    "url": "https://github.com/ipython/ipython/blob/56e6925dfa50e2c7f4a6471547b8176275db7c25/IPython/utils/_process_win32.py#L20"
                },
                {
                    "url": "https://github.com/advisories/GHSA-29gw-9793-fvw7"
                }
            ],
            "published_at": "2023-02-10T19:55:53Z",
            "updated_at": "2023-09-02T05:03:05Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "IPython"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": "< 8.10",
                    "first_patched_version": {
                        "identifier": "8.10"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:L/AC:H/PR:N/UI:R/S:U/C:L/I:L/A:L",
                "score": 4.5
            },
            "cwes": [
                {
                    "cwe_id": "CWE-20",
                    "name": "Improper Input Validation"
                },
                {
                    "cwe_id": "CWE-78",
                    "name": "Improper Neutralization of Special Elements used in an OS Command ('OS Command Injection')"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "IPython"
            },
            "severity": "medium",
            "vulnerable_version_range": "< 8.10",
            "first_patched_version": {
                "identifier": "8.10"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/5",
        "html_url": "https://github.com/username/gradio/security/dependabot/5",
        "created_at": "2024-05-24T01:34:12Z",
        "updated_at": "2024-05-24T01:34:12Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 4,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "certifi"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-43fp-rhv2-5gv8",
            "cve_id": "CVE-2022-23491",
            "summary": "Certifi removing TrustCor root certificate",
            "description": "Certifi 2022.12.07 removes root certificates from \"TrustCor\" from the root store. These are in the process of being removed from Mozilla's trust store.\n\nTrustCor's root certificates are being removed pursuant to an investigation prompted by media reporting that TrustCor's ownership also operated a business that produced spyware. Conclusions of Mozilla's investigation can be found [here](https://groups.google.com/a/mozilla.org/g/dev-security-policy/c/oxX69KFvsm4/m/yLohoVqtCgAJ).",
            "severity": "medium",
            "identifiers": [
                {
                    "value": "GHSA-43fp-rhv2-5gv8",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2022-23491",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://github.com/certifi/python-certifi/security/advisories/GHSA-43fp-rhv2-5gv8"
                },
                {
                    "url": "https://github.com/certifi/python-certifi/commit/9e9e840925d7b8e76c76fdac1fab7e6e88c1c3b8"
                },
                {
                    "url": "https://groups.google.com/a/mozilla.org/g/dev-security-policy/c/oxX69KFvsm4/m/yLohoVqtCgAJ"
                },
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2022-23491"
                },
                {
                    "url": "https://github.com/advisories/GHSA-43fp-rhv2-5gv8"
                }
            ],
            "published_at": "2022-12-07T23:05:18Z",
            "updated_at": "2023-01-29T05:05:37Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "certifi"
                    },
                    "severity": "medium",
                    "vulnerable_version_range": ">= 2017.11.05, < 2022.12.07",
                    "first_patched_version": {
                        "identifier": "2022.12.07"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:H/UI:N/S:C/C:N/I:H/A:N",
                "score": 6.8
            },
            "cwes": [
                {
                    "cwe_id": "CWE-345",
                    "name": "Insufficient Verification of Data Authenticity"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "certifi"
            },
            "severity": "medium",
            "vulnerable_version_range": ">= 2017.11.05, < 2022.12.07",
            "first_patched_version": {
                "identifier": "2022.12.07"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/4",
        "html_url": "https://github.com/username/gradio/security/dependabot/4",
        "created_at": "2024-05-24T01:34:12Z",
        "updated_at": "2024-05-24T01:34:12Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 3,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "pillow"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-q4mp-jvh2-76fj",
            "cve_id": "CVE-2022-45199",
            "summary": "Pillow subject to DoS via SAMPLESPERPIXEL tag",
            "description": "Pillow starting with 9.2.0 and prior to 9.3.0 allows denial of service via SAMPLESPERPIXEL. A large value in the SAMPLESPERPIXEL tag could lead to a memory and runtime DOS in TiffImagePlugin.py when setting up the context for image decoding. This issue has been patched in version 9.3.0.",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-q4mp-jvh2-76fj",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2022-45199",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2022-45199"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/pull/6700"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/commit/2444cddab2f83f28687c7c20871574acbb6dbcf3"
                },
                {
                    "url": "https://bugs.gentoo.org/878769"
                },
                {
                    "url": "https://github.com/python-pillow/Pillow/releases/tag/9.3.0"
                },
                {
                    "url": "https://github.com/pypa/advisory-database/tree/main/vulns/pillow/PYSEC-2022-42980.yaml"
                },
                {
                    "url": "https://security.gentoo.org/glsa/202211-10"
                },
                {
                    "url": "https://github.com/advisories/GHSA-q4mp-jvh2-76fj"
                }
            ],
            "published_at": "2022-11-14T12:00:15Z",
            "updated_at": "2023-09-02T05:03:04Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "pillow"
                    },
                    "severity": "high",
                    "vulnerable_version_range": ">= 9.2.0, < 9.3.0",
                    "first_patched_version": {
                        "identifier": "9.3.0"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
                "score": 7.5
            },
            "cwes": [
                {
                    "cwe_id": "CWE-400",
                    "name": "Uncontrolled Resource Consumption"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "pillow"
            },
            "severity": "high",
            "vulnerable_version_range": ">= 9.2.0, < 9.3.0",
            "first_patched_version": {
                "identifier": "9.3.0"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/3",
        "html_url": "https://github.com/username/gradio/security/dependabot/3",
        "created_at": "2024-05-24T01:34:12Z",
        "updated_at": "2024-05-24T01:34:12Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 2,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "py"
            },
            "manifest_path": "test/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-w596-4wvx-j9j6",
            "cve_id": "CVE-2022-42969",
            "summary": "ReDoS in py library when used with subversion ",
            "description": "The py library through 1.11.0 for Python allows remote attackers to conduct a ReDoS (Regular expression Denial of Service) attack via a Subversion repository with crafted info data, because the InfoSvnCommand argument is mishandled.\n\nThe particular codepath in question is the regular expression at `py._path.svnurl.InfoSvnCommand.lspattern` and is only relevant when dealing with subversion (svn) projects. Notably the codepath is not used in the popular pytest project. The developers of the pytest package have released version `7.2.0` which removes their dependency on `py`. Users of `pytest` seeing alerts relating to this advisory may update to version `7.2.0` of `pytest` to resolve this issue. See https://github.com/pytest-dev/py/issues/287#issuecomment-1290407715 for additional context.",
            "severity": "high",
            "identifiers": [
                {
                    "value": "GHSA-w596-4wvx-j9j6",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2022-42969",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2022-42969"
                },
                {
                    "url": "https://github.com/pytest-dev/py/issues/287"
                },
                {
                    "url": "https://github.com/pytest-dev/py/blob/cb87a83960523a2367d0f19226a73aed4ce4291d/py/_path/svnurl.py#L316"
                },
                {
                    "url": "https://github.com/pytest-dev/py/issues/288"
                },
                {
                    "url": "https://github.com/pytest-dev/pytest/issues/10392"
                },
                {
                    "url": "https://news.ycombinator.com/item?id=34163710"
                },
                {
                    "url": "https://pypi.org/project/py"
                },
                {
                    "url": "https://github.com/advisories/GHSA-w596-4wvx-j9j6"
                }
            ],
            "published_at": "2022-10-16T12:00:23Z",
            "updated_at": "2023-08-17T05:01:50Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "py"
                    },
                    "severity": "high",
                    "vulnerable_version_range": "<= 1.11.0",
                    "first_patched_version": null
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
                "score": 7.5
            },
            "cwes": [
                {
                    "cwe_id": "CWE-1333",
                    "name": "Inefficient Regular Expression Complexity"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "py"
            },
            "severity": "high",
            "vulnerable_version_range": "<= 1.11.0",
            "first_patched_version": null
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/2",
        "html_url": "https://github.com/username/gradio/security/dependabot/2",
        "created_at": "2024-05-24T01:34:11Z",
        "updated_at": "2024-05-24T01:34:11Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    },
    {
        "number": 1,
        "state": "open",
        "dependency": {
            "package": {
                "ecosystem": "pip",
                "name": "torch"
            },
            "manifest_path": "demo/musical_instrument_identification/requirements.txt",
            "scope": "runtime"
        },
        "security_advisory": {
            "ghsa_id": "GHSA-47fc-vmwq-366v",
            "cve_id": "CVE-2022-45907",
            "summary": "PyTorch vulnerable to arbitrary code execution",
            "description": "In PyTorch before trunk/89695, torch.jit.annotations.parse_type_line can cause arbitrary code execution because eval is used unsafely. The fix for this issue is available in version 1.13.1. There is a release checker in [issue #89855](https://github.com/pytorch/pytorch/issues/89855).",
            "severity": "critical",
            "identifiers": [
                {
                    "value": "GHSA-47fc-vmwq-366v",
                    "type": "GHSA"
                },
                {
                    "value": "CVE-2022-45907",
                    "type": "CVE"
                }
            ],
            "references": [
                {
                    "url": "https://nvd.nist.gov/vuln/detail/CVE-2022-45907"
                },
                {
                    "url": "https://github.com/pytorch/pytorch/issues/88868"
                },
                {
                    "url": "https://github.com/pytorch/pytorch/commit/767f6aa49fe20a2766b9843d01e3b7f7793df6a3"
                },
                {
                    "url": "https://github.com/pytorch/pytorch/issues/89855"
                },
                {
                    "url": "https://github.com/pytorch/pytorch/pull/89189"
                },
                {
                    "url": "https://github.com/pytorch/pytorch/releases/tag/v1.13.1"
                },
                {
                    "url": "https://github.com/advisories/GHSA-47fc-vmwq-366v"
                }
            ],
            "published_at": "2022-11-26T03:30:27Z",
            "updated_at": "2023-01-30T05:01:22Z",
            "withdrawn_at": null,
            "vulnerabilities": [
                {
                    "package": {
                        "ecosystem": "pip",
                        "name": "torch"
                    },
                    "severity": "critical",
                    "vulnerable_version_range": "<= 1.13.0",
                    "first_patched_version": {
                        "identifier": "1.13.1"
                    }
                }
            ],
            "cvss": {
                "vector_string": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
                "score": 9.8
            },
            "cwes": [
                {
                    "cwe_id": "CWE-77",
                    "name": "Improper Neutralization of Special Elements used in a Command ('Command Injection')"
                }
            ]
        },
        "security_vulnerability": {
            "package": {
                "ecosystem": "pip",
                "name": "torch"
            },
            "severity": "critical",
            "vulnerable_version_range": "<= 1.13.0",
            "first_patched_version": {
                "identifier": "1.13.1"
            }
        },
        "url": "https://api.github.com/repos/username/gradio/dependabot/alerts/1",
        "html_url": "https://github.com/username/gradio/security/dependabot/1",
        "created_at": "2024-05-24T01:34:11Z",
        "updated_at": "2024-05-24T01:34:11Z",
        "dismissed_at": null,
        "dismissed_by": null,
        "dismissed_reason": null,
        "dismissed_comment": null,
        "fixed_at": null,
        "auto_dismissed_at": null
    }
]